src/ext/simple-package.lisp

```diff
@@ -0,0 +1,206 @@
+  (:documentation "It downloads the SOURCE to the desired location."))
+
+(defvar *git-base-arglist* (list "git")
+  "The git program, to be appended command-line options.")
+
+(defun run-git (arglist)
+  (uiop:wait-process
+   (uiop:launch-program (concatenate 'list *git-base-arglist* arglist)
+                        :ignore-error-status t)))
+
+(defstruct (git (:include source)) url branch commit)
+
+(defmethod download-source ((source git) (output-location String))
+  (let ((output-dir (str:concat
+                     (namestring *packages-directory*) output-location)))
+    (run-git (list "clone" (git-url source) output-dir))
+    (when (git-branch source)
+      (uiop:with-current-directory (output-dir)
+        (run-git (list "checkout" "-b" (git-branch source)))))
+
+    (when (git-commit source)
+      (uiop:with-current-directory (output-dir)
+        (run-git (list "checkout" (git-commit source)))))
+    output-dir))
+
+(defstruct (quicklisp (:include source)))
+
+(eval-when (:compile-toplevel)
+  (defparameter *quicklisp-system-list*
+    (remove-duplicates
+     (mapcar #'ql-dist:release (ql:system-list)))))
+
+(defmethod download-source ((source quicklisp) (output-location String))
+  (let* ((output-dir (str:concat
+                     (namestring *packages-directory*) output-location))
+         (release (find (source-name source)
+                       *quicklisp-system-list*
+                       :key #'ql-dist:project-name
+                       :test #'string=))
+         (url (ql-dist:archive-url release))
+         (name (source-name source))
+         (tgzfile (str:concat name ".tgz"))
+         (tarfile (str:concat name ".tar")))
+    (if release
+        (prog1 output-dir
+          (uiop:with-current-directory (*packages-directory*)
+            (quicklisp-client::maybe-fetch-gzipped url tgzfile
+                                                   :quietly t)
+            (ql-gunzipper:gunzip tgzfile tarfile)
+            (ql-minitar:unpack-tarball tarfile)
+            (delete-file tgzfile)
+            (delete-file tarfile)
+            (uiop/cl:rename-file (ql-dist:prefix release) output-location)))
+        (editor-error "Package ~a not found!." (source-name source)))))
+
+(defmethod download-source (source output-location)
+  (editor-error "Source ~a not available." source))
+
+(defclass simple-package ()
+  ((name :initarg :name
+         :accessor simple-package-name)
+   (source :initarg :source
+           :accessor simple-package-source)
+   (directory :initarg :directory
+              :accessor simple-package-directory)))
+
+(defgeneric package-remove (package))
+
+(defmethod package-remove ((package simple-package))
+  (uiop:delete-directory-tree
+   (uiop:truename* (simple-package-directory package)) :validate t)
+  (delete package *installed-packages*))
+
+(defun packages-list ()
+  (remove-duplicates
+   (mapcar #'(lambda (d) (pathname (directory-namestring d)))
```

Comment:
I'm pretty sure `#'` in front of `lambda` is old style and not necessary.

Reply:
Indeed, I think this was some GNU Emacs flashback :laughing: